### PR TITLE
BOOKKEEPER-889: Bookie client does not try to use unhealthy bookies w…

### PR DESCRIPTION
**Pull request by sboobna**
…hen forming ensembles

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcher.java [hunk 1]
```diff
@@ -235,7 +255,15 @@ public void processResult(int rc, String path, Object ctx, List<String> children
      */
     public ArrayList<BookieSocketAddress> newEnsemble(int ensembleSize, int writeQuorumSize)
             throws BKNotEnoughBookiesException {
-        return placementPolicy.newEnsemble(ensembleSize, writeQuorumSize, EMPTY_SET);
+        try {
+            // we try to only get from the healthy bookies first
+            return placementPolicy.newEnsemble(ensembleSize, writeQuorumSize, quarantinedBookies.asMap().keySet());
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcher.java]: the key set will be modified in #newEnsembel(). It seems to me you have to create a new set from current quarantinedBookies's key set

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcher.java [hunk 2]
```diff
@@ -250,7 +278,30 @@ public void processResult(int rc, String path, Object ctx, List<String> children
     public BookieSocketAddress replaceBookie(List<BookieSocketAddress> existingBookies, int bookieIdx)
             throws BKNotEnoughBookiesException {
         BookieSocketAddress addr = existingBookies.get(bookieIdx);
-        return placementPolicy.replaceBookie(addr, new HashSet<BookieSocketAddress>(existingBookies));
+        try {
+            // we exclude the quarantined bookies also first
+            List<BookieSocketAddress> existingAndQuarantinedBookies = Lists.newArrayList();
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcher.java]: Why not create a hash set directly?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestBookieHealthCheck.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    public TestBookieHealthCheck() {
+        super(4);
+        baseClientConf.setAddEntryTimeout(1);
+        baseClientConf.enableBookieHealthCheck();
+        baseClientConf.setBookieHealthCheckInterval(1, TimeUnit.SECONDS);
+        baseClientConf.setBookieErrorThresholdPerInterval(1);
+        baseClientConf.setBookieQuarantineTime(5, TimeUnit.SECONDS);
+    }
+
+    @Test
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestBookieHealthCheck.java]: @Test(timeout = 60000)

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestBookieHealthCheck.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        Assert.assertTrue(lh3.getLedgerMetadata().getEnsemble(0).contains(bookieToQuarantine));
+
+        // make sure faulty bookie is out of quarantine
+        Thread.sleep(baseClientConf.getBookieQuarantineTimeSeconds() * 1000);
+
+        // the bookie should not be quarantined anymore
+        Assert.assertFalse(bkc.bookieWatcher.quarantinedBookies.asMap().containsKey(bookieToQuarantine));
+    }
+
+    @Test
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestBookieHealthCheck.java]: @Test(timeout = 60000)

**Comment by sboobna** [comment]
@sijie updated
I have also added a few more bk errors which we do not consider for it to be quarantined.

**Comment by sijie** [comment]
the new change looks good to me. +1

**Pull request closed without merging**